clippy_lints/src/unbound_return_lifetimes.rs

```diff
@@ -0,0 +1,128 @@
+use if_chain::if_chain;
+use rustc::declare_lint_pass;
+use rustc::hir::*;
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc_session::declare_tool_lint;
+
+use crate::utils::span_lint;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for lifetime annotations on return values
+    /// which might not always get bound.
+    ///
+    /// **Why is this bad?** If the function contains unsafe code which
+    /// transmutes to the lifetime, the resulting lifetime may live longer
+    /// than was intended.
+    ///
+    /// **Known problems*: None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// // Bad: unbound return lifetime causing unsoundnes
+    /// fn foo<'a>(x: impl AsRef<str> + 'a) -> &'a str {
+    ///    let s = x.as_ref();
+    ///    unsafe { &*(s as *const str) }
+    /// }
+    /// // Good: bound return lifetime is sound
+    /// struct WrappedStr(str);
+    /// fn foo<'a>(x: &'a str) -> &'a WrappedStr {
+    ///   unsafe { &*(x as *const str as *const WrappedStr) }
+    /// }
```

Comment:
Could you move `WrappedStr` above the `Bad` comment and use it in bad function as return type for more symmetry between the two examples? Also could you please separate the `}` and `//` with a newline?

Reply:
Done.

Reply:
Maybe also say that when the bad foo takes a `String`, `'a` can be equal to `'static`?

Reply:
Added an example note.